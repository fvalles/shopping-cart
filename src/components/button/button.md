Basic button component with default background='primary' and text color='white':

```jsx
<Button title="Press me" onClick={() => null} />
```

Disabled button component:

```jsx
<Button disabled title="Press me" onClick={() => null} />
```

Button component with custom background and text color:

```jsx
<Button
  background="white"
  color="primary"
  title="Press me"
  onClick={() => null}
/>
```
