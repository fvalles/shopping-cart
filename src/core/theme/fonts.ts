export const Fonts = {
  light: 'Avenir Light',
  medium: 'Avenir Medium',
  regular: 'Avenir Regular',
};

export type FontsType = typeof Fonts;
