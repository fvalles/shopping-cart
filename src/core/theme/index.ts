export { Colors } from './colors';
export { Fonts } from './fonts';
export { GlobalStyle } from './global-style';
