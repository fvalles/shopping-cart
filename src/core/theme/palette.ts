export const palette = {
  // color names are based on https://colornamer.netlify.app/
  black: '#000000',
  blackMarket: '#212240',
  blackMarketLight: '#21224029',
  bleachedSilk: '#F3F3F3',
  faintFern: '#DBDBE0',
  mischka: '#A6A7B3',
  sharpGrey: '#CACAD1',
  stormGrey: '#717285',
  venetianNights: '#7350FF',
  white: '#FFFFFF',
};
