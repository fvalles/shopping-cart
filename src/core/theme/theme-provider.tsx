import React, { FunctionComponent } from 'react';
import { DefaultTheme, ThemeProvider as Provider } from 'styled-components';
import { Colors, Fonts } from '.';

interface ThemeProviderProps {
  children?: React.ReactNode;
}

export const ThemeProvider: FunctionComponent<ThemeProviderProps> = ({
  children,
}) => (
  <Provider
    theme={
      {
        Colors,
        Fonts,
      } as DefaultTheme
    }>
    {children}
  </Provider>
);
