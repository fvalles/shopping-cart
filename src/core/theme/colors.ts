import { palette } from './palette';

export const Colors = {
  // color names are based on https://colornamer.netlify.app/
  primary: palette.venetianNights,
  productImageBorder: palette.sharpGrey,
  productPrice: palette.black,
  productQuantity: palette.faintFern,
  productTitle: palette.mischka,
  summaryBackground: palette.bleachedSilk,
  summaryHeading: palette.stormGrey,
  summaryText: palette.blackMarket,
  white: palette.white,
};

export type ColorsType = typeof Colors;
export type KeyColors = keyof ColorsType;
