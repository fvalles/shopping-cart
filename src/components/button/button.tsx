import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { KeyColors } from '../../core/theme/colors';
import { ButtonText } from '../typography';

/**
 * Styled Components
 */

export const StyledButton = styled.button<StyledButtonProps>`
  background-color: ${({ background = 'primary', theme }) =>
    theme.Colors[background]};
  border-radius: 4px;
  padding-bottom: 16px;
  padding-top: 16px;
  width: 100%;
`;

/**
 * Types
 */

interface StyledButtonProps {
  background?: KeyColors;
}

interface ButtonProps {
  /** Text to be displayed inside the button */
  title: string;
  /** Action to execute when the button is clicked */
  onClick: () => void;
  /** Button background color */
  background?: KeyColors;
  /** Button title color */
  color?: KeyColors;
}

/**
 * Button
 */

export const Button: FunctionComponent<ButtonProps> = ({
  title,
  onClick,
  background,
  color,
}) => (
  <StyledButton background={background} onClick={onClick}>
    <ButtonText color={color}>{title}</ButtonText>
  </StyledButton>
);
