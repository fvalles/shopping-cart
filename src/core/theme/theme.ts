import { DefaultTheme } from 'styled-components';
import { Colors } from './colors';

const theme: DefaultTheme = {
  Colors,
};

export { theme };
