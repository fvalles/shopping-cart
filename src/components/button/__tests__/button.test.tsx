import React from 'react';
import { ThemeProvider } from '@core/theme/theme-provider';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { Button } from '../button';

describe('button component', () => {
  it('should render with title, onClick function and default background and text colors', () => {
    const result = renderer
      .create(
        <ThemeProvider>
          <Button
            onClick={jest.fn()}
            testId="some-test-id"
            title="Some Title"
          />
        </ThemeProvider>,
      )
      .toJSON();

    expect(result).toHaveStyleRule('background-color', '#7350FF');
    expect(result).toHaveStyleRule('border-radius', '4px');
    expect(result).toHaveStyleRule('padding-bottom', '16px');
    expect(result).toHaveStyleRule('padding-top', '16px');
    expect(result).toHaveStyleRule('width', '100%');

    const buttonText = result.children[0];
    expect(buttonText).toHaveStyleRule('color', '#FFFFFF');
    expect(result).toMatchSnapshot();
  });

  it('should render with title, onClick function and custom primary background and white text', () => {
    const result = renderer
      .create(
        <ThemeProvider>
          <Button
            background="white"
            color="primary"
            onClick={jest.fn()}
            testId="some-test-id"
            title="Some Title"
          />
        </ThemeProvider>,
      )
      .toJSON();

    expect(result).toHaveStyleRule('background-color', '#FFFFFF');
    expect(result).toHaveStyleRule('border-radius', '4px');
    expect(result).toHaveStyleRule('padding-bottom', '16px');
    expect(result).toHaveStyleRule('padding-top', '16px');
    expect(result).toHaveStyleRule('width', '100%');

    const buttonText = result.children[0];
    expect(buttonText).toHaveStyleRule('color', '#7350FF');
    expect(result).toMatchSnapshot();
  });
});
