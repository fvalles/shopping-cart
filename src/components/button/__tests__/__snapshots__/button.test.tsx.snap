// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button component should render with title, onClick function and custom primary background and white text 1`] = `
.c1 {
  color: #7350FF;
  font-family: Avenir Regular,'Avenir Next','Segoe UI', sans-serif;
  font-size: 16px;
  font-weight: bold;
  line-height: 14px;
}

.c0 {
  background-color: #FFFFFF;
  border-radius: 4px;
  padding-bottom: 16px;
  padding-top: 16px;
  width: 100%;
}

<button
  className="c0"
  onClick={[MockFunction]}
>
  <span
    className="c1"
    color="primary"
  >
    Some Title
  </span>
</button>
`;

exports[`button component should render with title, onClick function and default background and text colors 1`] = `
.c1 {
  color: #FFFFFF;
  font-family: Avenir Regular,'Avenir Next','Segoe UI', sans-serif;
  font-size: 16px;
  font-weight: bold;
  line-height: 14px;
}

.c0 {
  background-color: #7350FF;
  border-radius: 4px;
  padding-bottom: 16px;
  padding-top: 16px;
  width: 100%;
}

<button
  className="c0"
  onClick={[MockFunction]}
>
  <span
    className="c1"
  >
    Some Title
  </span>
</button>
`;
